File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.py
from collections import deque
def solution(maps):
    N, M = len(maps), len(maps[0])
    visited = [[0]*M for _ in range(N)]

    answer = []
    for i in range(N):
        for j in range(M):
            if maps[i][j]=='X' or visited[i][j]:
                continue
            queue = deque()
            queue.append((i,j))
            visited[i][j]=1
            n_food = int(maps[i][j])
            while queue:
                i0, j0 = queue.popleft()
                for di, dj in [(1,0), (0,1), (-1,0), (0,-1)]:
                    ni, nj = i0+di, j0+dj
                    if 0<=ni<N and 0<=nj<M and not visited[ni][nj] and maps[ni][nj]!='X':
                        queue.append((ni,nj))
                        visited[ni][nj] = 1
                        n_food += int(maps[ni][nj])
            answer.append(n_food)
    if not answer:
        answer.append(-1)
    else:
        answer.sort()
    return answer

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.js
function solution(maps) {
  const answer = [];
  let [row, col] = [maps.length, maps[0].length];
  let [dx, dy] = [
    [-1, 0, 1, 0],
    [0, -1, 0, 1],
  ];
  const table = maps.map((v) => v.split(""));

  for (let i = 0; i < row; i++) {
    for (let j = 0; j < col; j++) {
      if (table[i][j] === "X") continue;
      let cnt = +table[i][j];
      table[i][j] = "X";
      let stack = [[i, j]];

      while (stack.length) {
        let [x, y] = stack.pop();
        for (let k = 0; k < 4; k++) {
          let [nx, ny] = [x + dx[k], y + dy[k]];
          if (
            nx >= 0 &&
            nx < row &&
            ny >= 0 &&
            ny < col &&
            table[nx][ny] !== "X"
          ) {
            cnt += +table[nx][ny];
            table[nx][ny] = "X";
            stack.push([nx, ny]);
          }
        }
      }
      answer.push(cnt);
    }
  }
  answer.sort((a, b) => a - b);
  return answer.length ? answer : [-1];
}